core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory2.java

```diff
@@ -0,0 +1,248 @@
+
+    @VisibleForTesting
+    static final Attributes.Key<AtomicReference<ConnectivityStateInfo>> STATE_INFO =
+        Attributes.Key.of("state-info");
+
+    public RoundRobinLoadBalancer(Helper helper) {
+      this.helper = helper;
+    }
+
+    @Override
+    public void handleResolvedAddresses(List<ResolvedServerInfoGroup> servers,
+        Attributes attributes) {
+      Set<EquivalentAddressGroup> currentAddrs = ImmutableSet.copyOf(subchannels.keySet());
+      Set<EquivalentAddressGroup> latestAddrs =
+          resolvedServerInfoGroupToEquivalentAddressGroup(servers);
+      Set<EquivalentAddressGroup> addedAddrs = Sets.difference(latestAddrs, currentAddrs);
+      Set<EquivalentAddressGroup> removedAddrs = Sets.difference(currentAddrs, latestAddrs);
+
+      Attributes subchannelAttrs = Attributes.newBuilder()
+          .set(STATE_INFO, new AtomicReference<ConnectivityStateInfo>(
+              ConnectivityStateInfo.forNonError(IDLE)))
+          .build();
+
+      // Create new subchannels for new addresses.
+      for (EquivalentAddressGroup addressGroup : addedAddrs) {
+        Subchannel subchannel = checkNotNull(helper.createSubchannel(addressGroup, subchannelAttrs),
+            "subchannel");
+        subchannels.put(addressGroup, subchannel);
+        subchannel.requestConnection();
+      }
+
+      // Shutdown subchannels for removed addresses.
+      for (EquivalentAddressGroup addressGroup : removedAddrs) {
+        Subchannel subchannel = subchannels.remove(addressGroup);
+        subchannel.shutdown();
+      }
+
+      updatePicker(null);
+    }
+
+    @Override
+    public void handleNameResolutionError(Status error) {
+      updatePicker(error);
+    }
+
+    @Override
+    public void handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+      if (!subchannels.inverse().containsKey(subchannel)) {
+        return;
+      }
+      if (stateInfo.getState() == IDLE) {
+        subchannel.requestConnection();
+      }
+      getSubchannelStateInfoRef(subchannel).set(stateInfo);
+      updatePicker(getAggregatedError());
+    }
+
+    @Override
+    public void shutdown() {
+      for (Subchannel subchannel : getSubchannels()) {
+        subchannel.shutdown();
+      }
+    }
+
+    /**
+     * Updates picker with the list of active subchannels (state != TRANSIENT_ERROR).
+     */
+    private void updatePicker(@Nullable Status error) {
+      List<Subchannel> activeList = filterFailingSubchannels(getSubchannels());
+      helper.updatePicker(new Picker(activeList, error));
+    }
+
+    /**
+     * Filters out failing subchannels (state == TRANSIENT_ERROR).
+     */
+    private static List<Subchannel> filterFailingSubchannels(Collection<Subchannel> subchannels) {
+      List<Subchannel> nonFailingSubchannels = Lists.newArrayList();
+      for (Subchannel subchannel : subchannels) {
+        if (getSubchannelStateInfoRef(subchannel).get().getState() != TRANSIENT_FAILURE) {
```

Comment:
The managed channel can only use subchannels that are READY. If IDLE or CONNECTING subchannels are picked, the RPC will be returned to the managed channel's buffer, and will be processed when a new picker is provided.

This would (unintentionally) produce a funny result, if, for example, there is 1 READY subchannel and 9 IDLE or CONNECTING subchannels. 1 out of 10 RPCs will be assigned to that READY subchannel and proceed, while the rest 90% will be buffered. If the state later becomes 2 READY and 8 CONNECTING, a picker will be created and all the buffered RPCs will face a 20% chance of proceed.

That doesn't seem to be a reasonable behavior.  It would make more sense to only return READY subchannels.